File: app/src/main/java/com/nedaluof/qurany/ui/component/RecitersAdapter.java
package com.nedaluof.qurany.ui.component;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.nedaluof.qurany.data.DataManager;
import com.nedaluof.qurany.data.model.Reciter;
import com.nedaluof.qurany.databinding.ItemReciterBinding;
import com.nedaluof.qurany.ui.reciter.ReciterView;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import dagger.hilt.android.qualifiers.ApplicationContext;

/**
 * Created by nedaluof on 7/12/2020.
 */
public class RecitersAdapter extends RecyclerView.Adapter<RecitersAdapter.RecitersViewHolder> {
    private ReciterView onClickHandler;
    private List<Reciter> list = new ArrayList<>();
    private Context context;
    private DataManager dataManager;

    @Inject
    public RecitersAdapter(DataManager dataManager, @ApplicationContext Context context) {
        this.dataManager = dataManager;
        this.context = context;
    }

    //must set before set data list to the adapter
    public void setOnClickHandler(ReciterView onClickHandler) {
        this.onClickHandler = onClickHandler;
    }

    @NonNull
    @Override
    public RecitersViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new RecitersViewHolder(ItemReciterBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false).getRoot());
    }

    @Override
    public void onBindViewHolder(@NonNull RecitersViewHolder holder, int position) {
        holder.binding.tvReciterName.setText(list.get(position).getName().toUpperCase());
        holder.binding.tvReciterRewaya.setText(list.get(position).getRewaya().toUpperCase());
        holder.binding.tvReciterSuraCount.setText("There ".concat(list.get(position).getCount()).concat(" Sura"));
        holder.binding.reciterDataLayout.setOnClickListener(v -> onClickHandler.onClickGetReciterData(list.get(position)));
        if (list.get(position).getInMyReciters()) {
            holder.binding.imgAddFavorite.setVisibility(View.GONE);
        } else {
            holder.binding.imgAddFavorite.setVisibility(View.VISIBLE);
        }

        holder.binding.imgAddFavorite.setOnClickListener(v -> {
            onClickHandler.onClickAddToMyReciters(list.get(position));
            holder.binding.imgAddFavorite.setVisibility(View.GONE);
        });
    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    public void add(Reciter reciter) {
        list.add(reciter);
        notifyItemInserted(list.size() - 1);
    }

    public void addAll(List<Reciter> list) {
        for (Reciter reciter : list) {
            add(reciter);
        }
    }

    public void remove(Reciter reciter) {
        int position = list.indexOf(reciter);
        if (position > -1) {
            list.remove(position);
            notifyItemRemoved(position);
        }
    }

    public void clear() {
        while (getItemCount() > 0) {
            remove(getItem(0));
        }
    }

    public Reciter getItem(int position) {
        return list.get(position);
    }

    static class RecitersViewHolder extends RecyclerView.ViewHolder {
        private ItemReciterBinding binding;

        public RecitersViewHolder(@NonNull View view) {
            super(view);
            binding = ItemReciterBinding.bind(view);
        }
    }
}

File: app/src/main/java/com/nedaluof/qurany/ui/component/RecitersAdapter.kt
package com.nedaluof.qurany.ui.component

import android.content.Context
import android.os.Build
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable
import androidx.annotation.RequiresApi
import androidx.recyclerview.widget.RecyclerView
import com.nedaluof.qurany.R
import com.nedaluof.qurany.data.model.Reciter
import com.nedaluof.qurany.databinding.ItemReciterBinding
import com.nedaluof.qurany.util.Utility


/**
 * Created by nedaluof on 12/11/2020.
 */
class RecitersAdapter(
        context: Context
) : RecyclerView.Adapter<RecitersAdapter.RecitersVH>(),
        Filterable {

    lateinit var listener: ReciterAdapterListener
    private val recitersData = ArrayList<Reciter>()
    private val filterRecitersData = ArrayList<Reciter>()
    private val surasCountLabel = context.getString(R.string.sura_count)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecitersVH {
        val binding = ItemReciterBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return RecitersVH(binding)
    }

    override fun onBindViewHolder(holder: RecitersVH, position: Int) {
        val reciter = recitersData[position]
        with(holder.binding) {
            tvReciterName.text = reciter.name
            tvReciterRewaya.text = reciter.rewaya
            val surasCount = "$surasCountLabel ${reciter.count}"
            tvReciterSuraCount.text = surasCount
            reciterDataLayout.setOnClickListener { listener.onReciterClicked(reciter) }

            if (reciter.inMyReciters) {
                imgFavorite.visibility = View.GONE
            } else {
                imgFavorite.visibility = View.VISIBLE
            }

            imgFavorite.setOnClickListener {
                listener.onAddToFavoriteClicked(reciter)
                //need good solution
                //todo update this and progress from ReciterFragment
                imgFavorite.visibility = View.GONE
            }
        }
    }

    override fun getItemCount(): Int = recitersData.size

    interface ReciterAdapterListener {
        fun onReciterClicked(reciter: Reciter)
        fun onAddToFavoriteClicked(reciter: Reciter)
    }

    fun addReciters(reciters: ArrayList<Reciter>) {
        recitersData.apply {
            clear()
            addAll(reciters)
        }

        notifyDataSetChanged()

        filterRecitersData.apply {
            clear()
            addAll(reciters)
        }
    }

    override fun getFilter(): Filter = searchFilter


    private val searchFilter = object : Filter() {
        @RequiresApi(Build.VERSION_CODES.N)
        override fun performFiltering(constraint: CharSequence?): FilterResults {
            val filterResults = FilterResults()
            filterResults.values = if (constraint == null || constraint.isEmpty()) {
                filterRecitersData
            } else {
                if (Utility.getLanguage() == "_arabic") {
                    val pattern = constraint.toString().trim { it <= ' ' }
                    filterRecitersData.filter { reciter ->
                        reciter.name?.contains(pattern)!!
                    }
                } else {
                    filterRecitersData.filter { reciter ->
                        reciter.name?.firstName()?.contains(constraint.toString().trim { it <= ' ' })!! ||
                                reciter.name?.secondName()?.contains(constraint.toString())!!
                    }
                }
            }
            return filterResults
        }

        override fun publishResults(p0: CharSequence?, results: FilterResults?) {
            with(recitersData) {
                clear()
                addAll(results?.values as List<Reciter>)
            }
            notifyDataSetChanged()
        }

    }

    fun String.firstName(): String = this.substring(0, this.indexOf(' '))
    fun String.secondName(): String = this.substring(this.indexOf(' ') + 1)

    class RecitersVH(val binding: ItemReciterBinding) : RecyclerView.ViewHolder(binding.root)
}